README.md
<h1 align="center">Welcome to log-filesize-webpack-plugin 👋</h1>

<p>
  <a href="https://www.npmjs.com/package/@jsany/log-filesize-webpack-plugin" target="_blank">
    <img alt="npm version" src="https://img.shields.io/npm/v/@jsany/log-filesize-webpack-plugin/latest?logo=npm&style=flat-square" />
  </a>
  <a href="https://www.npmjs.com/package/@jsany/log-filesize-webpack-plugin" target="_blank">
    <img alt="npm miniziped size" src="https://img.shields.io/bundlephobia/minzip/@jsany/log-filesize-webpack-plugin/latest?logo=npm&style=flat-square" />
  </a>
  <a href="https://www.npmjs.com/package/@jsany/log-filesize-webpack-plugin" target="_blank">
    <img alt="node version" src="https://img.shields.io/node/v/@jsany/log-filesize-webpack-plugin?color=blue&style=flat-square" />
  </a>
  <a href="https://www.npmjs.com/package/@jsany/log-filesize-webpack-plugin" target="_blank">
    <img alt="GitHub Workflow Status" src="https://img.shields.io/github/workflow/status/jsany/log-filesize-webpack-plugin/Publish?logo=github&style=flat-square" />
  </a>
  <a href="https://codecov.io/gh/jsany/log-filesize-webpack-plugin" target="_blank">
    <img alt="codecov" src="https://img.shields.io/codecov/c/github/jsany/log-filesize-webpack-plugin/main?logo=codecov&style=flat-square" />
  </a>
  <a href="https://github.com/jsany/log-filesize-webpack-plugin#readme" target="_blank">
    <img alt="documentation" src="https://img.shields.io/badge/documentation-yes-brightgreen.svg?style=flat-square" />
  </a>
  <a href="https://github.com/jsany/log-filesize-webpack-plugin/graphs/commit-activity" target="_blank">
    <img alt="Maintained" src="https://img.shields.io/badge/Maintained%3F-yes-green.svg?style=flat-square" />
  </a>
  <a href="https://github.com/jsany/log-filesize-webpack-plugin/blob/main/LICENSE" target="_blank">
    <img alt="License" src="https://img.shields.io/badge/License-MIT-yellow.svg?style=flat-square" />
  </a>
</p>

> log assets size pretty after webpack build(wirte by typescript)

### 🏠 [Homepage](https://github.com/jsany/log-filesize-webpack-plugin#readme)

### ✨ [Example](https://github.com/jsany/log-filesize-example)

<https://github.com/jsany/log-filesize-example/blob/main/webpack.config.js#L168>

## Prerequisites

- node >=10
- webpack >=4

## Install

```sh
yarn add -D @jsany/log-filesize-webpack-plugin
# or npm install --save-dev @jsany/log-filesize-webpack-plugin
```

## Setup

### In `webpack.config.js`:

```js
const LogFilesizeWebpackPlugin = require('@jsany/log-filesize-webpack-plugin');

module.exports = {
  stats: 'none', // to close webpack print
  // ...
  plugins: [
    new LogFilesizeWebpackPlugin()
    // or
    // new LogFilesizeWebpackPlugin({gzip: true,...OtherOptions})
  ]
  // ...
};
```

### Then run build, it will look like this:

![example](screenshots/example.png)

## Options

| Options    | Type                                                                                               | Default                                                     | Description                                                           |
| ---------- | -------------------------------------------------------------------------------------------------- | ----------------------------------------------------------- | --------------------------------------------------------------------- |
| `gzip`     | boolean                                                                                            | `true`                                                      | Print gzipped filesize                                                |
| `errors`   | boolean                                                                                            | `true`                                                      | Print compile errors                                                  |
| `warnings` | boolean                                                                                            | `true`                                                      | Print compile warnings                                                |
| `hash`     | boolean                                                                                            | `true`                                                      | Print current compile hash                                            |
| `version`  | boolean                                                                                            | `true`                                                      | Print current webpack version                                         |
| `time`     | boolean                                                                                            | `true`                                                      | Print compile cost time                                               |
| `builtAt`  | boolean                                                                                            | `true`                                                      | Print when to start compiling                                         |
| `maxSize`  | number                                                                                             | `1.8 * 1024 * 1024` (Byte)                                  | These sizes are pretty large. We'll warn for bundles exceeding them   |
| `priority` | { js?: number, css?: number,image?: number,font?: number,json?: number,txt?: number,html?: number} | `{ js: 10, css: 9,image: 8,font: 7,json: 6,txt: 5,html: 4}` | The larger the value, the higher the sort, and hide if less than zero |

## Feature

- [x] Support javascript and switch log it
- [x] Support css and switch log it
- [x] Support image and switch log it
- [x] Support json and switch log it
- [x] Support font and switch log it
- [x] Support txt and switch log it
- [x] Support html and switch log it
- [x] Support custom assets log priority

## Todo

- [ ] Support custom file color
- [ ] ...

### Support files with these extensions: [ext look detail](./src/helper/mimeMap.ts)

## Author

👤 **jiangzhiguo2010**

- Website: <https://github.com/Mr-jiangzhiguo/book>
- Github: [@jsany](https://github.com/jsany)

## 🤝 Contributing

Contributions, issues and feature requests are welcome!<br />Feel free to check [issues page](https://github.com/jsany/log-filesize-webpack-plugin/issues). You can also take a look at the [contributing guide](https://github.com/jsany/log-filesize-webpack-plugin/blob/main/CONTRIBUTING.md).

## Show your support

Give a ⭐️ if this project helped you!

## 📝 License

Copyright © 2020 [jiangzhiguo2010](https://github.com/jsany).<br />
This project is [MIT](https://github.com/jsany/log-filesize-webpack-plugin/blob/main/LICENSE) licensed.

---

_This README was generated with ❤️ by [readme-md-generator](https://github.com/kefranabg/readme-md-generator)_


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.2.13](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.12...v1.2.13) (2020-09-11)

### [1.2.12](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.11...v1.2.12) (2020-08-19)

### Bug Fixes

- transform path \*.d.ts alias([#4](https://github.com/jsany/log-filesize-webpack-plugin/issues/4)) ([996dbd7](https://github.com/jsany/log-filesize-webpack-plugin/commit/996dbd77c7b6e82f8d839a46e366e4bfedd856d6))

### [1.2.11](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.10...v1.2.11) (2020-08-06)

### Bug Fixes

- cancel ci push tags ([564ca95](https://github.com/jsany/log-filesize-webpack-plugin/commit/564ca95bbe5b1b2280d6c2039ff45998becb0bc8))

### [1.2.10](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.8...v1.2.10) (2020-08-01)

### Bug Fixes

- gitub workflow and update doc ([b7ace8c](https://github.com/jsany/log-filesize-webpack-plugin/commit/b7ace8ccdc518e3e77297b6bf8f1924f0fc2f7cb))
- push action add git config ([be63eff](https://github.com/jsany/log-filesize-webpack-plugin/commit/be63eff69aa1feb9d8e4d09468faa9570e0e714c))
- workflow/push-action branch apply to main ([0fc134c](https://github.com/jsany/log-filesize-webpack-plugin/commit/0fc134c4b1bd48d9e22d990521eacfa7c3e6569e))

### [1.2.9](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.8...v1.2.9) (2020-08-01)

### Bug Fixes

- gitub workflow and update doc ([b7ace8c](https://github.com/jsany/log-filesize-webpack-plugin/commit/b7ace8ccdc518e3e77297b6bf8f1924f0fc2f7cb))
- push action add git config ([be63eff](https://github.com/jsany/log-filesize-webpack-plugin/commit/be63eff69aa1feb9d8e4d09468faa9570e0e714c))
- workflow/push-action branch apply to main ([0fc134c](https://github.com/jsany/log-filesize-webpack-plugin/commit/0fc134c4b1bd48d9e22d990521eacfa7c3e6569e))

### [1.2.8](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.7...v1.2.8) (2020-07-24)

### Bug Fixes

- github actions workflow optimize ([c7a1126](https://github.com/jsany/log-filesize-webpack-plugin/commit/c7a1126a05985169ffd0e04ff4dbcf5dc05b363c))

### [1.2.7](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.6...v1.2.7) (2020-07-21)

### Bug Fixes

- add badge ([39ab065](https://github.com/jsany/log-filesize-webpack-plugin/commit/39ab0658f494dca6c8eb3101e19ae7f805907f65))

### [1.2.6](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.5...v1.2.6) (2020-07-21)

### Bug Fixes

- workflows test rm node 8 ([81a0749](https://github.com/jsany/log-filesize-webpack-plugin/commit/81a074972ad36480fd3326d6c6660902c6d56dca))

### [1.2.5](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.4...v1.2.5) (2020-07-21)

### Bug Fixes

- workflows/publish env (it only use in with and name) ([d133cfa](https://github.com/jsany/log-filesize-webpack-plugin/commit/d133cfaa603fcecc2a73908f6ffc0045171ac683))

### [1.2.4](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.3...v1.2.4) (2020-07-21)

### Bug Fixes

- workflows/publish jobs.coverage.env ([9cb31cc](https://github.com/jsany/log-filesize-webpack-plugin/commit/9cb31ccfdfbaab1436d4d66bde08e1dd9db10305))

### [1.2.3](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.2...v1.2.3) (2020-07-21)

### Bug Fixes

- use persional token to trigger event(https://github.com/actions/create-release/issues/52) ([6253fab](https://github.com/jsany/log-filesize-webpack-plugin/commit/6253fabcfa3adf307e03e43f16cbef7ee1253d1e))

### [1.2.2](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.1...v1.2.2) (2020-07-21)

### Bug Fixes

- optimize github workflows ([a858fed](https://github.com/jsany/log-filesize-webpack-plugin/commit/a858fedea1c163d5daea951d0a4dd41f72781544))

### [1.2.1](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.2.0...v1.2.1) (2020-07-21)

### Bug Fixes

- github workflow ([bfde043](https://github.com/jsany/log-filesize-webpack-plugin/commit/bfde043fca1318db381911affe25c31cbc960673))
- github workflows/npm-publish ([e609cb0](https://github.com/jsany/log-filesize-webpack-plugin/commit/e609cb066e7ba07d8181a1842d21bdbb2bbb2a85))

## [1.2.0](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.1.5...v1.2.0) (2020-07-20)

### Features

- support Github Actions([#3](https://github.com/jsany/log-filesize-webpack-plugin/issues/3)) ([4f3c668](https://github.com/jsany/log-filesize-webpack-plugin/commit/4f3c668e3132e0de264a9d3af57e5f7470612ed6))

### [1.1.5](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.1.4...v1.1.5) (2020-07-20)

### [1.1.4](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.1.3...v1.1.4) (2020-07-20)

### [1.1.3](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.1.2...v1.1.3) (2020-07-18)

### Bug Fixes

- support unitTest ([#2](https://github.com/jsany/log-filesize-webpack-plugin/issues/2)) ([b8020f8](https://github.com/jsany/log-filesize-webpack-plugin/commit/b8020f81d2dc55a1e65e5c00007e26ce1a69c9d1))

### [1.1.2](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.1.1...v1.1.2) (2020-07-17)

### [1.1.1](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.1.0...v1.1.1) (2020-07-17)

## [1.1.0](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.0.6...v1.1.0) (2020-07-17)

### Features

- assets support image, font, json, and txt ([5883e38](https://github.com/jsany/log-filesize-webpack-plugin/commit/5883e383c407c7f88b182a172ccabcf6124f2d95))
- optimize example ([00251ee](https://github.com/jsany/log-filesize-webpack-plugin/commit/00251eeca262419295e64d06a88c804bbba3467d))

### [1.0.6](https://github.com/jsany/log-filesize-webpack-plugin/compare/v1.0.5...v1.0.6) (2020-07-16)


src/index.ts
import { Compiler, Stats } from 'webpack';
import { join, relative } from 'path';
import chalk from 'chalk';
import filesize from '@/helper/filesize';
import getGzippedSize from '@/helper/getGzippedSize';
import makeRow from '@/helper/makeRow';
const ui = require('cliui')({ with: 80 });
import mime from '@/helper/definedMime';
import { TKnownType, knownTypes } from '@/helper/mimeMap';

export type TPriority = Partial<Record<TKnownType, number>>;
export interface IOptions {
  /* Switch log gzipped filesize */
  gzip?: boolean;
  /* Switch log compile errors */
  errors?: boolean;
  /* Switch log compile warnings */
  warnings?: boolean;
  /* Switch log current compile hash */
  hash?: boolean;
  /* Switch log current webpack version */
  version?: boolean;
  /* Switch log compile cost time */
  time?: boolean;
  /* Switch log when to start compiling */
  builtAt?: boolean;
  /* These js sizes are pretty large. We'll warn for bundles exceeding them. */
  maxSize: number;
  /* The larger the value, the higher the sort, and hide if less than zero */
  priority?: TPriority;
}
export interface IStatus {
  msg: 'success' | 'error' | 'norun';
  data: any;
}

export interface IAssetMeta {
  name: string;
  type: TKnownType;
  size: number;

  gzipped?: number;

  suggested?: boolean;
  priority?: number;
}

class LogFilesizeWebpackPlugin {
  public static defaultOptions: IOptions = {
    gzip: true,
    errors: true,
    warnings: true,
    hash: true,
    version: true,
    time: true,
    builtAt: true,
    maxSize: 1.8 * 1024 * 1024,
    priority: {
      js: 10,
      css: 9,
      image: 8,
      font: 7,
      json: 6,
      txt: 5,
      html: 4
    }
  };
  public options: IOptions;

  public status: IStatus;
  public assetsCache: IAssetMeta[];

  public constructor(options = {} as IOptions) {
    this.options = this.mergeOptions(options, LogFilesizeWebpackPlugin.defaultOptions);
    this.status = { msg: 'norun', data: 'printfStats havent run' };
    this.assetsCache = [];
  }
  public apply = (compiler: Compiler) => {
    compiler.hooks.afterEmit.tap('LogFilesizeWebpackPlugin', (compilation) => {
      // DONE: use file.type(MIME) instead of regexp
      // console.log('===== compilation =====');
      const { assets } = compilation;
      const seenNames = new Map();
      this.assetsCache = Object.keys(assets)
        .filter((item) => {
          if (seenNames.has(item)) {
            return false;
          }
          seenNames.set(item, true);
          const type = mime.getType(item) as TKnownType;
          const priority = this.options.priority?.[type];
          return knownTypes.includes(type) && priority! > 0;
        })
        .map((item) => {
          const source = assets[item].source();
          const name = item;
          const type = mime.getType(name) as TKnownType;
          const size = assets[item].size() as number;
          const gzipped = (this.options.gzip && getGzippedSize(source)) || undefined;
          const priority = this.options.priority?.[type];
          const isLarge = size > this.options.maxSize;

          return { name, type, size, gzipped, priority, suggested: isLarge && type === 'js' };
        })
        .sort((a, b) => {
          if (a.type === b.type) {
            return b.size - a.size;
          } else {
            return b.priority! - a.priority!;
          }
        });
      // console.log('===== compilation =====');
    });
    compiler.hooks.done.tap('LogFilesizeWebpackPlugin', (stats) => {
      // console.log('===== done =====');
      // // console.log(stats);
      // // console.log(this.assetsCache)
      // console.log('===== done =====');
      this.printfStats(stats);
    });
  };

  public printfStats = (stats: Stats) => {
    try {
      const json = stats.toJson({
        modules: false,
        chunks: false
      });
      // console.log('===== printfStats =====');
      // console.log(json);
      // console.log('===== printfStats =====');
      const dir = relative(process.cwd(), json.outputPath || '');

      if (this.options.errors && json.errors.length) {
        ui.div(chalk.red.bold('Error: ') + '\n' + chalk.red(json.errors.join('\n')));
        ui.div('\n');
      }
      if (this.options.warnings && json.warnings.length) {
        ui.div(chalk.yellow.bold('Warning: ') + '\n' + chalk.yellow(json.warnings.join('\n')));
        ui.div('\n');
      }
      if (this.options.version) {
        ui.div(chalk.white.bold('Version: ') + chalk.white('webpack ' + json.version));
      }
      if (this.options.hash) {
        ui.div(chalk.white.bold('Hash: ') + chalk.white(json.hash));
      }
      if (this.options.time) {
        ui.div(chalk.white.bold('Time: ') + chalk.white(json.time + 'ms'));
      }
      if (this.options.builtAt) {
        ui.div(chalk.white.bold('Built At: ') + chalk.white(new Date(json.builtAt!)));
      }
      ui.div('\n');
      ui.div(
        makeRow(
          chalk.cyan.bold(`File`),
          chalk.cyan.bold(`Size`),
          (this.options.gzip && chalk.cyan.bold(`Gzipped`)) || ''
        ) +
          `\n\n` +
          this.assetsCache
            ?.map((asset) => {
              let row1 = join(dir, asset.name);
              const row2 = filesize(asset.size);
              const row3 = (this.options.gzip && filesize(asset.gzipped!)) || '';
              switch (asset.type) {
                case 'js':
                  row1 = asset.suggested ? chalk.yellow(row1) : chalk.green(row1);
                  break;
                case 'css':
                  row1 = chalk.blue(row1);
                  break;
                case 'image':
                  row1 = chalk.magenta(row1);
                  break;
                case 'font':
                  row1 = chalk.cyan(row1);
                  break;
                case 'json':
                  row1 = chalk.gray(row1);
                  break;
                case 'txt':
                  row1 = chalk.white(row1);
                  break;
                case 'html':
                  row1 = chalk.red(row1);
                  break;
                default:
                  break;
              }
              return makeRow(row1, row2, row3);
            })
            .join(`\n`)
      );
      console.log(
        `${ui.toString()}\n\n  ${chalk.gray.inverse(
          `🚨 Assets other than ${knownTypes.slice(0, -1).join(', ')} and ${knownTypes.slice(
            -1
          )} are omitted.`
        )}\n`
      );
      if (this.assetsCache?.some((asset) => asset.suggested)) {
        console.log(chalk.yellow('The bundle size is significantly larger than recommended.'));
        console.log(
          chalk.yellow(
            'Consider reducing it with code splitting: https://v4.webpack.docschina.org/guides/code-splitting/'
          )
        );
      }
      this.status = { msg: 'success', data: this.assetsCache };
    } catch (err) {
      this.status = { msg: 'error', data: err };
      throw new Error(err);
    }
  };

  public mergeOptions = (options: IOptions, defaults: IOptions): IOptions => {
    try {
      const { priority: optionsPriority, ...optionsRest } = options;
      const { priority: defaultPriority, ...defaultRest } = defaults;

      Object.assign(defaultPriority, optionsPriority);
      Object.assign(defaultRest, optionsRest, { priority: defaultPriority });
      // console.log('===== defaultRest =====');
      // console.log(defaultRest)
      // console.log('===== defaultRest =====');
      return defaultRest;
    } catch (err) {
      throw new Error(err);
    }
  };
}

// to generate type definitions
export default LogFilesizeWebpackPlugin;
// to support CommonJS
module.exports = LogFilesizeWebpackPlugin;
// to support ES6 default import
module.exports.default = module.exports;
